newdocs/typeclass-cheatsheet.md

```diff
@@ -0,0 +1,223 @@
+ class (GetPrevHash b, ConsensusProtocol p) => BlockSupportsProtocol b where              
+   validateView :: bc → Header b → ValidateView p  -- project from hdr for hdr validation
+   selectView   :: bc → Header b → SelectView p    -- project from hdr for chain selection
+```
+Classes connected to ledgers:
+```haskell
+  class GetTip l where                         
+    getTip :: l → Point l               -- Point of the most recently applied block
+
+  class (GetTip l, GetTip (Ticked l)) => IsLedger l where
+    type family LedgerErr l      :: Type                   
+    type family AuxLedgerEvent l :: Type
+    applyChainTickLedgerResult   :: lc → SlotNo → l → LedgerResult l (Ticked l)  -- apply slot based state transformations (tip unchanged)
+        
+  class (IsLedger l, HeaderHash l ~ HeaderHash b, HasHeader b, HasHeader (Header b)) => ApplyBlock l b where
+    applyBlockLedgerResult   :: lc → b → Ticked l → Except (LedgerErr l) (LedgerResult l l)  
+    reapplyBlockLedgerResult :: lc → b → Ticked l →                       LedgerResult l l
+    
+  class ApplyBlock (LedgerState b) b => UpdateLedger b where
+    {}
+
+  -- | Link protocol to ledger
+  class (BlockSupportsProtocol b, UpdateLedger b, ValidateEnvelope b) => LedgerSupportsProtocol b where
+    protocolLedgerView   :: lc → Ticked l → Ticked ledvw   -- 'ledvw' ('LedgerView (BlockProtocol b)') extracted from the ledger
+    ledgerViewForecastAt :: lc → l → Forecast ledvw        -- get a forecast (of future 'ledvw's) from a given ledger state.
+      
+  class (UpdateLedger b) => LedgerSupportsMempool b where
+    txInvariant :: GenTx b → Bool                                                -- check if internal invariants of the transaction hold
+    applyTx   :: lc → WhetherToIntervene → SlotNo → tx → tls → Except txerr (tls, Validated tx)      -- apply an unvalidated transaction
+    reapplyTx :: lc →            SlotNo → Validated tx → tls → Except txerr tls          -- apply a previously validated transaction ...
+    txsMaxBytes :: tls → Word32                                   -- max number of bytes of transactions that can be put into a block    
+    txInBlockSize :: tx → Word32                                  -- post-serialisation size in bytes of a 'GenTx b'
+    txForgetValidated :: Validated tx → tx                        -- discard the evidence that transaction has been previously validated     
+```
+
+## Quick Reference for Useful Base Types
+### ouroboros-consensus:
+
+**ouroboros-consensus/src/Ouroboros/Consensus/Forecast.hs`**
+
+``` haskell
+-- | Forecast the effect of time ticking
+data Forecast a = Forecast {
+      forecastAt  :: WithOrigin SlotNo
+
+      -- Precondition: @At s >= forecastAt@
+    , forecastFor :: SlotNo -> Except OutsideForecastRange (Ticked a)
+    }
+```
+
+**ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Basics.hs**
+
+``` haskell
+-- | The result of invoke a ledger function that does validation
```

Comment:
Probably our typo, copy-pasted, but might as well fix it: `of invoking`.

Can probably drop the `deriving` clause here.

Reply:
done